Improve top level parallelization algorithm and fix no make target bug

nolanrdavenport opened the pull request:
Maximizes the amount of packages being built simultaneously at any given time, effectively doubling the amount of time savings you get from using br-pbuilder. Created workaround for the no make target bug. Modified pb_finalize_targets to only call target-post-image because it includes all of the other finalize steps. Before, it would execute some finalize steps multiple times.

.gitignore: Added build objects.

The idea behind the new algorithm is to start the build for a package as soon as two conditions are met:
1. All of the package's dependencies were successfully built.
2. There is at least 1 thread available to create for the package to use for it's build.

From my testing, I used the beaglebone_qt5 configuration. With the previous br-pbuilder algorithm, I had 12% build time savings, and with the new algorithm, I had 23% build time savings.

Note that this was using this hardware:
Ryzen 9 3900X 12 core 24 thread
NVME SSD
32GB DDR4 RAM

For the no make target bug, a workaround was created that detects this error condition and just retries the build. From my testing, I no longer have build errors as a result of this. It's a workaround which isn't as ideal as finding a solution using lock files, but since this is the only solution we have available, I think it's worth merging to increase the viability of using br-pbuilder at least until we can find a proper solution.

paguilar commented:
The algorithm improvements seem quite nice, thanks for contributing!

I was aware that the current algorithm that works by following the priority levels has some drawbacks specifically when some packages of the next level could start building, but they're put on hold until the last package of the current level finishes.
This improvement covers those gaps reducing even more the overall building time.
I'm testing its functionality and its performance; the former seems to be as robust as before, while the latter is the real improvement that I'm still measuring.

As far as I've seen, this commit contains 3 changes that address 3 different issues: The algorithm improvement, the .gitignore file and the 'no make target' workaround.
I would kindly ask you if you could split it in 3 different commit requests or patches, so each single commit addresses a single issue. In this way is easier to debug and understand in the long term what changed in a given commit.
I would like to avoid to have a single commit that has a nice improvement along with a workaround for a bug.

Thanks again!

nolanrdavenport committed:
.gitignore: Added br-pbuilder build objects that do not need to be
comitted to the remote repository.

Signed-off-by: Nolan Davenport <nolanrdavenport@gmail.com>

nolanrdavenport committed:
graph_exec.c: Inside of pb_finalize_targets, these three targets were removed: host-finalize, staging-finalize, and target-finalize. The reason is because target-post-image already performs these steps so they aren't needed, and by including them, they were performed multiple times. In my case, I noticed that files were copied from configured rootfs_overlays multiple times instead of once.

Signed-off-by: Nolan Davenport <nolanrdavenport@gmail.com>

nolanrdavenport committed:
graph_exec.c: Added workaround for an error where sometimes make isn't
able to find the make target for a package, causing the build to fail.
This happened randomly and only affected a percentage of builds. This solution
is just to get br-pbuilder working for now, but in the future, this
issue should be debugged.

Signed-off-by: Nolan Davenport <nolanrdavenport@gmail.com>

nolanrdavenport committed:
graph_exec.c: Changed the top-level parallelization algorithm to one
that maximizes the amount of packages being built in parallel at any
given time. This further increases time savings.

The idea behind the new algorithm is to start the build for a package as soon as two conditions are met (give or take a second):
1. All of the package's dependencies were successfully built.
2. There is at least 1 thread available to create for the package to use for it's build.

I used the beaglebone_qt5 configuration for testing. With the previous br-pbuilder algorithm, I had 12% build time savings, and with the new algorithm, I had 23% build time savings.

The above test was performed with this hardware:
Ryzen 9 3900X 12 core 24 thread
NVME SSD
32GB DDR4 RAM

Signed-off-by: Nolan Davenport <nolanrdavenport@gmail.com>

nolanrdavenport commented:
> The algorithm improvements seem quite nice, thanks for contributing!
> 
> I was aware that the current algorithm that works by following the priority levels has some drawbacks specifically when some packages of the next level could start building, but they're put on hold until the last package of the current level finishes. This improvement covers those gaps reducing even more the overall building time. I'm testing its functionality and its performance; the former seems to be as robust as before, while the latter is the real improvement that I'm still measuring.
> 
> As far as I've seen, this commit contains 3 changes that address 3 different issues: The algorithm improvement, the .gitignore file and the 'no make target' workaround. I would kindly ask you if you could split it in 3 different commit requests or patches, so each single commit addresses a single issue. In this way is easier to debug and understand in the long term what changed in a given commit. I would like to avoid to have a single commit that has a nice improvement along with a workaround for a bug.
> 
> Thanks again!

Sounds good, I went ahead and split this PR into multiple commits!

paguilar commented:
I cherry-picked the four commits and format their commit message. Now the threads_pool branch contains these changes.
I'll keep testing and measuring the improved algorithm and at the same time trying to find a proper fix for the _'no make target'_ issue.
   
Thanks a lot!

paguilar reviewed (commented):
I cherry-picked the four commits and format their commit message. Now the threads_pool branch contains these changes.
It was not done following Github's recommended flow, but next time I could follow it.
I think this PR could be closed.
